ow_faults_injection/src/FaultInjector.cpp

```diff
@@ -0,0 +1,206 @@
+  if (m_joint_state_indices.empty()) {
+    for (int j = 0; j < NUM_JOINTS; j ++) {
+      int index = findPositionInGroup(msg->name, joint_names[j]);
+      if (index >= 0)
+        m_joint_state_indices.push_back(index);
+    }
+  }
+
+  for (int i = 0; i < flag_msg.flags.size(); i ++){
+    flag_msg.flags[i] = 0.0;
+  }
+
+  // Set failed sensor values to 0
+  unsigned int index;
+
+  // checking rqt
+  checkCamFaults();
+  const float SET_FLAG = 1;
+  //ant faults
+  if (m_faults.ant_pan_encoder_failure || m_faults.ant_pan_effort_failure 
+      && findJointIndex(J_ANT_PAN, index)){
+    flag_msg.flags[index] = SET_FLAG;
+  }
+  if ( m_faults.ant_tilt_encoder_failure || m_faults.ant_tilt_effort_failure 
+      && findJointIndex(J_ANT_TILT, index)) {
+    flag_msg.flags[index] = SET_FLAG;
+  }
+
+  //arm faults
+  if (m_faults.shou_yaw_encoder_failure && findJointIndex(J_SHOU_YAW, index)) {
+    output.position[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+  if (m_faults.shou_yaw_effort_failure && findJointIndex(J_SHOU_YAW, index)) {
+    output.effort[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+
+  if (m_faults.shou_pitch_encoder_failure && findJointIndex(J_SHOU_PITCH, index)) {
+    output.position[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+  if (m_faults.shou_pitch_effort_failure && findJointIndex(J_SHOU_PITCH, index)) {
+    output.effort[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+
+  if (m_faults.prox_pitch_encoder_failure && findJointIndex(J_PROX_PITCH, index)) {
+    output.position[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+  if (m_faults.prox_pitch_effort_failure && findJointIndex(J_PROX_PITCH, index)) {
+    output.effort[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+
+  if (m_faults.dist_pitch_encoder_failure && findJointIndex(J_DIST_PITCH, index)) {
+    output.position[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+  if (m_faults.dist_pitch_effort_failure && findJointIndex(J_DIST_PITCH, index)) {
+    output.effort[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+
+  if (m_faults.hand_yaw_encoder_failure && findJointIndex(J_HAND_YAW, index)) {
+    output.position[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+  if (m_faults.hand_yaw_effort_failure && findJointIndex(J_HAND_YAW, index)) {
+    output.effort[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+
+  if (m_faults.scoop_yaw_encoder_failure && findJointIndex(J_SCOOP_YAW, index)) {
+    output.position[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
+  if (m_faults.scoop_yaw_effort_failure && findJointIndex(J_SCOOP_YAW, index)) {
+    output.effort[index]  = FAULT_ZERO_TELEMETRY;
+    flag_msg.flags[index] = SET_FLAG;
+  }
```

Comment:
This is not what or how I expected to implement the fault flags. You are double checking these fault flags.
I would rather add the fault_msg as part of the JointFaults. We are already populating the joint faults there.

Reply:
updating